include/irr/video/IGPUVirtualTexture.h

```diff
@@ -45,8 +45,8 @@ class IGPUVirtualTexture final : public asset::IVirtualTexture<video::IGPUImageV
             base_t(
                 createGPUImageFromCPU(_driver, _am, _cpuStorage->image.get()),
                 _cpuStorage->m_assignedPageTableLayers,
-                _cpuStorage->m_addr_layerShift,
-                _cpuStorage->m_addr_xMask
+                _cpuStorage->m_decodeAddr_layerShift,
+                _cpuStorage->m_decodeAddr_xMask
```

Comment:
these should be constexpr

Reply:
those are things for decoding address from pool allocator into 3d (x,y,layer) address in order to encode it into what goes into page table (which indeed has constexpr masks/shifts: 4bits X, 4bits Y, 8bits layer)

Reply:
didn't we fix the number of bits on that as well?

Reply:
i dont understand how it could be fixed. If i have `tilesPerDim`=4, then number of bits for x and y page-wise page coords is 2 (log2(tilesPerDim) in general). If i had those bitlengths fixed, i'd just be reading/decoding wrong values.
For example pool allocator gave me `0b00000101'11000011` address. So in case of `tilesPerDim=4` i'm getting (`0b11`, `0b00`, `0b1011100`) 3D address and in case of `tilesPerDim=8` i'm getting (`0b011`,`0b000`,`0b10111`). So it cant be fixed because decoding 1D->3D is directly dependent on tiles per dim

Reply:
well not if you're gonna use the addresses directly (without a little translation)

Btw do you require that the physical texture be PoT ? Cause we could use morton ordering for the pool allocations as well (and the physical textures "could" look prettier in RenderDoc)

Reply:
> Btw do you require that the physical texture be PoT ?

yes, tilesPerDim is PoT

> Cause we could use morton ordering for the pool allocations as well (and the physical textures "could" look prettier in RenderDoc)

complicates physical page allocation a lot (i have to pad needed pages count to PoT and after packing deallocate not used...) and "prettyness" depends on fragmentation. Plus fragmentation becomes much bigger problem because of deallocating not-used pages resulting from PoT-padding...

Reply:
> complicates physical page allocation a lot (i have to pad needed pages count to PoT and after packing deallocate not used...) and "prettyness" depends on fragmentation. Plus fragmentation becomes much bigger problem because of deallocating not-used pages resulting from PoT-padding...

You actually don't need to align or pad anything, the tiles are not connected in any way whatsoever (its a pool allocator, there is no fragmentation).

turning an address of a single tile into an XY coordinate is bit extraction, positioning them morton style would essentially introduce just a permutation on the bit pattern (just after `multi_alloc_addr` and `multi_free_addr` on the pool allocator)

---------------------------

Anyway this is going off on a tangent (whether to use morton or not is completely orthogonal to the below).

We can fix the allocation of bits as a constant. Using the following rationale.

Max physical texture size in hardware is 16k, but it would be best to limit the X,Y dimensions to a minimum because of IEEE754 precision issues near `vec2(1.0,1.0)`

Maximum number of 2D array layers is 256, which means that we can use 8 bits of layers, and then 4 each for for X and Y (which gives us a reasonable 2k up to 8k 2D physical storage dimension depending on page size).

all you need to do is just convert the pool allocator address to a fixed byte form
```cpp
struct SPhysPageAddr
{
    uint8_t xy; // low 4 bits x, high 4 bits y
    uint8_t layer;
};

SPhysPageAddr poolAddrToCoord(uint16_t addr);
uint16_t coordToPoolAddr(SPhysPageAddr coord);
```

Reply:
> all you need to do is just convert the pool allocator address to a fixed byte form

thats exactly what im doing